### PR TITLE
Updated and refactored urg code in school_stats

### PR DESCRIPTION
**Pull request by bakerfranke**
## Description

PR to show re-factor of urg calcs in school_stats.

I couldn't build due to a permissions error, but wanted to get the PR up to show differences for review with jordan.

### CODE REVIEW

File: dbt/models/staging/dashboard/stg_dashboard__school_stats_by_years.sql [hunk 1]
```diff
@@ -39,125 +47,91 @@ school_stats_by_years as (
...
-
-        --  total_urm_no_tr_students / total_students as pct_urm_no_tr,
-        -- total_urm_students / total_students as pct_urm
-        
+            ) then 1
+            else 0
+        end as is_stage_hi,
+
+        coalesce(ssby.students_total, 0) as students, -- create aliases for all race groups. Force 0 in case student count is null
+        coalesce(ssby.student_am_count, 0) as student_am,
```
Comment (jordan-springer) [inline dbt/models/staging/dashboard/stg_dashboard__school_stats_by_years.sql:93]: are you sure we want to use both `NULL` and 0 values? It looks like the data coming from the app has both...

File: dbt/models/marts/courses/dim_course_status.sql
Comment (jordan-springer) [inline dbt/models/marts/courses/dim_course_status.sql:1]: removed here, will be built in its own PR

File: dbt/models/staging/dashboard/base/base_dashboard__schools.sql [hunk 2]
```diff
@@ -20,6 +20,7 @@ renamed as (
         latitude,
         longitude,
         school_category,
+        state_school_id, -- (js) recently removed (see: https://github.com/code-dot-org/code-dot-org/pull/53871#event-10428323040)
```
Comment (jordan-springer) [inline dbt/models/staging/dashboard/base/base_dashboard__schools.sql:23]: update per developer update (see note)

File: dbt/models/staging/dashboard/stg_dashboard__school_stats_by_years.sql [hunk 3]
```diff
@@ -72,114 +140,15 @@ school_stats_by_years as (
                 '11',
                 '12'
              ) then 1 
+             when grades_offered_hi is null then null 
             else 0 
-        end                                                                             as is_stage_hi,
-
-        sum(count_student_am
```
Comment (bakerfranke) [inline dbt/models/staging/dashboard/stg_dashboard__school_stats_by_years.sql:78]: Noting to myself and other readers.  This logic was moved up into the 'combined' CTE, not just deleted entirely.

File: dbt/models/marts/schools/fct_school_stats.sql [hunk 4]
```diff
@@ -1,118 +1,103 @@
...
     from {{ ref('stg_dashboard__school_stats_by_years') }}
 ),
 
-school_districts as (
-    select * 
-    from {{ ref('stg_dashboard__school_districts')}}
-),
+combined as (
+    select *,
+        -- calculations
```
[the remark this answers is not in the document]
Comment (bakerfranke) [reply dbt/models/marts/schools/fct_school_stats.sql:9]: I mispoke a little ^^ there in terms of consistency with the historical school_stats table.  My original commit acknowledged and covered the cases that produce the fields we'd been using.  Let's look to that.

File: dbt/models/marts/schools/fct_school_stats.sql [hunk 5]
```diff
@@ -1,118 +1,103 @@
...
-    from {{ ref('stg_dashboard__school_districts')}}
-),
+combined as (
+    select *,
+        -- calculations 
+        sum(count_student_am 
+            + count_student_as
+            + count_student_hi
+            + count_student_bl
+            + count_student_wh
```
[the remark this answers is not in the document]
Comment (jordan-springer) [reply dbt/models/marts/schools/fct_school_stats.sql:14]: that doesn't match what is in production right now, so this would be a change to the definiton of URG?

File: dbt/models/marts/schools/fct_school_stats.sql [hunk 6]
```diff
@@ -1,118 +1,103 @@
...
 ),
 
-school_districts as (
-    select * 
-    from {{ ref('stg_dashboard__school_districts')}}
-),
+combined as (
+    select *,
+        -- calculations 
+        sum(count_student_am
```
[the remark this answers is not in the document]
Comment (jordan-springer) [reply dbt/models/marts/schools/fct_school_stats.sql:10]: consistency and to ensure grain, a redundancy that is like a safety-check

File: dbt/models/marts/schools/fct_school_stats.sql [hunk 7]
```diff
@@ -1,118 +1,103 @@
...
+        
+        case when total_students = total_students_calculated
+             then sum(count_student_am 
+                + count_student_hi  
+                + count_student_bl  
+                + count_student_hp) 
+                / total_students::float 
+        end as urg_percent,
+        
+        case when .7 <= sum(count_student_am
```
[the remark this answers is not in the document]
Comment (jordan-springer) [reply dbt/models/marts/schools/fct_school_stats.sql:41]: this is coming from `analysis.school_stats`, I agree it's unnecessary

File: dbt/models/marts/schools/fct_school_stats.sql [hunk 8]
```diff
@@ -1,118 +1,103 @@
...
+                + count_student_hp)
+                / 
+                sum(count_student_am
+                + count_student_as
+                + count_student_hi
+                + count_student_bl
+                + count_student_wh
+                + count_student_hp
+                + count_student_tr)::float 
+        end as urg_percent_true,
```
Comment (jordan-springer) [inline dbt/models/marts/schools/fct_school_stats.sql:61]: @bakerfranke what about the `urg_perfect_true` and calcs after? do those meet expectations for what you would want to see migrated?

File: dbt/models/marts/schools/fct_school_stats.sql [hunk 9]
```diff
@@ -1,118 +1,103 @@
...
-        ssby.urg_percent_no_tr,
+        sum(count_student_am  
+            + count_student_as  
+            + count_student_hi  
+            + count_student_bl  
+            + count_student_wh  
+            + count_student_hp
+            + count_student_tr) as total_students_calculated,
+        
+        case when total_students = total_students_calculated
```
[the remark this answers is not in the document]
Comment (jordan-springer) [reply dbt/models/marts/schools/fct_school_stats.sql:33]: me neither, safe to remove? @bakerfranke

**Review by jordan-springer — COMMENTED** [review]
@bakerfranke thanks for the feedback! I made the requested changes and added a few more questions/responses to your review. Ready for you to look at it again.

File: dbt/models/staging/dashboard/base/base_dashboard__schools.sql (hunk 2 above)
Comment (bakerfranke) [reply dbt/models/staging/dashboard/base/base_dashboard__schools.sql:23]: acknowledge.

File: dbt/models/marts/schools/fct_school_stats.sql [hunk 10]
```diff
@@ -1,118 +1,56 @@
 with 
+school_stats_by_years as (
+    select *,
 
-schools as (
```
Comment (bakerfranke) [inline dbt/models/marts/schools/fct_school_stats.sql:3]: This is nice and clean for the urg stuff.  Nice.

Overall question and something I think I missed in first review:  if `fct_school_stats.sql` is supposed to replace the existing analysis.school_stats then there are a bunch of fields missing.  You've got good calculations in here from the stuff that came from school_stats_by_years, but it's missing things like school address, state, zip, etc. that came from {{ ref('stg_dashboard__schools') }}. which it seems like you've removed?  

Additionally, analysis.school_stats does the work of pulling the information from the most recent entries in schools and school_stats by years, so it ends up being 1 record per distinct school_id.

Was this intentional?  am I missing something?

**Review by bakerfranke — COMMENTED** [review]
quibble/question.  wouldn't school_stats be a dimensional table not a fact table?  
Everything in school_stats is descriptive, or categorical information. We used to slice and dice the fact data like activity data going on at that school.

File: dbt/models/marts/schools/fct_school_stats.sql [hunk 11]
```diff
@@ -1,118 +1,103 @@
...
+            + count_student_hp
+            + count_student_tr) as total_students_calculated,
+        
+        case when total_students = total_students_calculated
+             then sum(count_student_am 
+                + count_student_hi  
+                + count_student_bl  
+                + count_student_hp) 
+                / total_students::float 
+        end as urg_percent,
```
[the remark this answers is not in the document]
Comment (bakerfranke) [reply dbt/models/marts/schools/fct_school_stats.sql]: urg is preferred.  We didn't change it in the past to keep legacy code working. Since these tables will potentially exposed for self-serve I agree now is the time to change.

File: dbt/models/marts/schools/fct_school_stats.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (bakerfranke) [reply dbt/models/marts/schools/fct_school_stats.sql]: let's remove then and use the variables we've now created.

File: dbt/models/marts/schools/fct_school_stats.sql (hunk 10 above)
Comment (jordan-springer) [reply dbt/models/marts/schools/fct_school_stats.sql:3]: Yes, I would consider those fields **dimensions** of schools (for dim_schools). In a fact table you traditionally have "facts" with FK's to tie out for those types of dimensions.
